packages/safe-ds-lang/src/language/generation/safe-ds-python-generator.ts

```diff
@@ -381,20 +394,180 @@ export class SafeDsPythonGenerator {
+        targetPlaceholder: SdsPlaceholder,
+        statementsWithEffect: SdsStatement[],
+    ): SdsStatement[] {
+        const neededPlaceholders = new Set<string>();
+        const neededStatements = new Set<SdsStatement>();
+        this.getAllPlaceholdersAndStatementsRelevantForPlaceholder(
+            targetPlaceholder,
+            statementsWithEffect,
+            neededPlaceholders,
+            neededStatements,
+        );
+        // Get all statements in sorted order
+        return Array.from(neededStatements.values()).sort((a, b) => a.$containerIndex! - b.$containerIndex!);
+    }
+
+    private getAllPlaceholdersAndStatementsRelevantForPlaceholder(
+        targetPlaceholder: SdsPlaceholder,
+        allStatements: SdsStatement[],
+        neededPlaceholders: Set<string>,
+        neededStatements: Set<SdsStatement>,
+    ) {
+        // If placeholder is inside a lambda, do not continue; Only placeholders inside the pipeline are collected
+        if (this.isNodeInsideBlockLambda(targetPlaceholder)) {
+            return;
+        }
+        // Find assignment of placeholder, to search used placeholders and impure dependencies
+
+        let assignment = getContainerOfType(targetPlaceholder, isSdsAssignment);
+        if (!assignment || !assignment.expression) {
+            /* c8 ignore next 2 */
+            return;
+        }
+        // If placeholder is already found, do not continue; It has been handled before
+        if (neededPlaceholders.has(targetPlaceholder.name) && neededStatements.has(assignment)) {
+            return;
+        }
+        neededPlaceholders.add(targetPlaceholder.name);
+        neededStatements.add(assignment);
+        // Get referenced placeholders of current placeholder. An expression in the assignment will always exist here
+        let placeholders = streamAllContents(assignment.expression!)
+            .filter(isSdsReference)
+            .filter((reference) => isSdsPlaceholder(reference.target.ref))
+            .map((reference) => <SdsPlaceholder>reference.target.ref!)
+            .toArray();
+        for (const referencedPlaceholder of placeholders) {
+            this.getAllPlaceholdersAndStatementsRelevantForPlaceholder(
+                referencedPlaceholder,
+                allStatements,
+                neededPlaceholders,
+                neededStatements,
+            );
+        }
+        // Get impure dependencies and follow them
+        let impureDependencies = this.getImpureDependenciesForStatement(assignment, allStatements);
+        for (const referencedStatement of impureDependencies) {
+            this.getAllPlaceholdersAndStatementsRelevantForStatement(
+                referencedStatement,
+                allStatements,
+                neededPlaceholders,
+                neededStatements,
+            );
+        }
+    }
+
+    private isNodeInsideBlockLambda(node: AstNode): boolean {
+        return Boolean(getContainerOfType(node, isSdsBlockLambda));
```

Comment:
Use `hasContainerOfType`. Then it's also doubtful whether this extra method is even need or whether it can be inlined.